indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/NumberEncoding.kt

```diff
@@ -241,222 +163,374 @@ fun Int.toUnsignedByteArray() = toLong().toUnsignedByteArray()
  *
- * @return the decoded unsigned BigInteger and the underlying varint-encoded bytes as `ByteArray`
+ * @return the decoded [BigInteger] and the underlying varint-encoded bytes as ByteString
  */
-inline fun Iterable<Byte>.decodeAsn1VarBigInt(): Pair<BigInteger, ByteArray> = iterator().decodeAsn1VarBigInt()
+fun Source.decodeAsn1VarBigInt():Pair<BigInteger, ByteArray> {
+    var result = BigInteger.ZERO
+    val accumulator = Buffer()
+    while (!exhausted()) {
+        val curByte = readByte()
+        val current = BigInteger(curByte.toUByte().toInt())
+        accumulator.writeByte(curByte)
+        result = (current and UVARINT_MASK_BIGINT) or (result shl 7)
+        if (current < UVARINT_SINGLEBYTE_MAXVALUE_UBYTE) break
+    }
+
+    return result to accumulator.readByteArray()
+}
+
 
 /**
- * Decodes an unsigned BigInteger from bytes using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
- * while the highest bit indicates if more bytes are to come. Trailing bytes are ignored.
+ * Writes a signed long using twos-complement encoding using the fewest bytes required
  *
- * @return the decoded unsigned BigInteger and the underlying varint-encoded bytes as `ByteArray`
+ * @return the number of byte written to the sink
  */
-inline fun ByteArray.decodeAsn1VarBigInt(): Pair<BigInteger, ByteArray> = iterator().decodeAsn1VarBigInt()
-
+fun Sink.writeTwosComplementLong(number: Long): Int = writeTwosComplement(number, true)
 
 /**
- * Decodes an BigInteger from bytes using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
- * while the highest bit indicates if more bytes are to come. Trailing bytes are ignored.
+ * Writes a signed long using twos-complement encoding using the fewest bytes required
+ * Allows for omitting the leading zero byte. This can be useful in certain number encodings such as
+ * in ASN.1 tag length encoding.
  *
- * @return the decoded BigInteger and the underlying varint-encoded bytes as `ByteArray`
- * @throws IllegalArgumentException if the number is larger than [ULong.MAX_VALUE]
+ * @return the number of byte written to the sink
  */
-fun Iterator<Byte>.decodeAsn1VarBigInt(): Pair<BigInteger, ByteArray> {
-    var result = BigInteger.ZERO
-    val mask = BigInteger.fromUByte(0x7Fu)
-    val accumulator = mutableListOf<Byte>()
-    while (hasNext()) {
-        val curByte = next()
-        val current = BigInteger(curByte.toUByte().toInt())
-        accumulator += curByte
-        result = (current and mask) or (result shl 7)
-        if (current < 0x80.toUByte()) break
+private fun Sink.writeTwosComplement(number: Long, padded: Boolean = true): Int = when {
+    (number >= -0x80L && number <= 0x7FL) -> {
+        writeByte(number.toByte())
+        1
+    }
+
+    (number >= -0x8000L && number <= 0x7FFFL) -> {
+        val byteWritten = writeOrSkipPadding(number, 8, padded)
```

Comment:
i do not see how this can ever write a leading zero byte. (also: please do not refer to a leading zero byte as "padding"; it is not padding.)

the matched `when` condition ensures that the number fits in 2 bytes without leading zero; leading zero is needed for range `0x8000` to `0xffff`

also why are we not writing the second byte? this entire function feels wrong the more i look at it

Reply:
reworked as per our offline discussion